src/segmenter/utils.rs

```diff
@@ -0,0 +1,68 @@
+use fst::raw::{Fst, Output};
+
+/// Final-state-transducer (FST) Segmenter
+pub(crate) struct FstSegmenter<'fst> {
+    words_fst: &'fst Fst<&'fst [u8]>
+}
+
+impl<'fst> FstSegmenter<'fst> {
+    pub(crate) fn new(words_fst: &'fst Fst<&'fst [u8]>) -> Self {
+        Self { words_fst }
+    }
+}
+
+impl<'fst> FstSegmenter<'fst> {
+    pub fn segment_str(&'fst self, mut to_segment: &'fst str) -> Box<dyn Iterator<Item = &'fst str> + 'fst> {
```

Comment:
Well, there is an issue here, `self` has the lifetime `'fst` which is good because it stores internally a reference to the dictionary. However `to_segment` is not related to the dictionary and so shouldn't have the `'fst` lifetime. But `to_segmentù needs a lifetime, this lifetime is the one related to the **original** text: `'o`, and because it is not needed for `FstSegmenter` but is needed for `segment_str` we will pass it locally:
```suggestion
    pub fn segment_str<'o>(&'fst self, mut to_segment: &'o str) -> Box<dyn Iterator<Item = &'o str> + 'o> {
```

Reply:
If we do that then we run into the same issue again, `'fst` may not outlive `'o`.

If I am not mistaken, using `'fst` for `to_segment` does not mean that it will have the same lifetime as the FST, but that the FST must have a longer lifetime than `to_segment`.

Based on my understanding of lifetimes, it is equivalent to:
```rs
pub fn segment_str<'o: 'fst>(&'fst self, mut to_segment: &'o str) -> Box<dyn Iterator<Item = &'o str> + 'fst> {
```
where `'o: 'fst` constraints `'o` to be shorter than `'fst`

Reply:
Yes you're right, but, I'd prefer to have an explicit `where 'fst: 'o` at the end of `segment_str` instead of putting the same name for every lifetime 😬

Reply:
Done! I learnt a lot about Rust from this, thanks!